### DISCUSSION TITLE
Many to many - allowing duplicates in junction table

### DISCUSSION THREAD

**Discussion by yann-mabsilico** [sqlalchemy/sqlalchemy#8599] [Usage Questions] [answered]
Hello,
I need to allow duplicates in a junction table. I have read https://github.com/sqlalchemy/sqlalchemy/discussions/8150 and the associated `joinedload` / `unique` combo, except in my case I actively want the duplicates to appear and am confused about how to proceed.

## Code
```
from sqlalchemy import Integer, Column, ForeignKey, Table
from sqlalchemy.orm import declarative_base, relationship, Session

Base = declarative_base()

class X(Base):
    index = Column(Integer, primary_key=True)
    ys = relationship('Y', secondary='XY')
    __tablename__ = 'X'
    
class Y(Base):
    index = Column(Integer, primary_key=True)
    __tablename__ = 'Y'
    
XY = Table(
        'XY',
        Base.metadata,
        Column('x_index', ForeignKey('X.index')),
        Column('y_index', ForeignKey('Y.index')),
)       

def insert():
    engine = create_engine(CONNECTION_STRING, echo=True, future=True)
    with Session(engine) as session:
        x = X(index=1)
        y = Y(index=10)
        x.ys = [y, y]
        session.merge(x)
        session.commit()

def select():
    engine = create_engine(CONNECTION_STRING, echo=True, future=True)
    with Session(engine) as session:
        x = X(index=1)
        x = session.merge(x)
        print(len(x.ys))

def insertUpdate():
    engine = create_engine(CONNECTION_STRING, echo=True, future=True)
    with Session(engine) as session:
        x = X(index=2)
        x.ys = [Y(index=20), Y(index=21)]
        x = session.merge(x)
        session.commit()
        
        x.ys[1] = x.ys[0]
        session.merge(x)
        session.commit()

def main():
    insert()
    select()
    insertUpdate()
    
main()
```

## Behavior and expectation
`insert` adds two rows in `XY`, both (1, 10). That is what I want.
`select` only finds one `Y` element (prints 1). I would want two.
In the `insertUpdate` function (an actual real case I need to handle), the first commit adds two rows (2, 20) and (2, 21) to `XY` while the second simply removes the (2, 21) row (and does not add a second (2, 20) row). I would want to have two (2, 20) rows.

**Comment by zzzeek** [comment]
> ## Behavior and expectation
> 
> `insert` adds two rows in `XY`, both (1, 10). 

the ORM won't do this unless you had an additional primary key in the association table and you used the [association object pattern](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object)/

> That is what I want. `select` only finds one `Y` element (prints 1). I would want two. 

the ORM doesn't support this nor is it really possible in general.

>  In the `insertUpdate` function (an actual real case I need to handle), the first commit adds two rows (2, 20) and (2, 21) to `XY` while the second simply removes the (2, 21) row (and does not add a second (2, 20) row). I would want to have two (2, 20) rows.

from a relational database perspective, there's no valid reason to have duplicate rows in a table (database relations are sets.  dupes are meaningless).  I'd ask what the actual problem you are trying to solve is, as proceeding using more traditional means will most definitely have a better end result.

**Reply by yann-mabsilico** [reply]
> > ## Behavior and expectation
> > `insert` adds two rows in `XY`, both (1, 10).
> 
> the ORM won't do this unless you had an additional primary key in the association table and you used the [association object pattern](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object)/

I'm not sure what you're saying. This is exactly what I'm observing: strictly using the code in the op and using the tables below (no primary key on `XY`), the `insert` function adds two rows to `XY`, both (1, 10). This is a postgres database, debian stable, sqlalchemy 1.4.39.
```
create table "X" (index integer primary key);
create table "Y" (index integer primary key);
create table "XY" (x_index integer, y_index integer);
alter table "XY" add foreign key (x_index) references "X" (index);
alter table "XY" add foreign key (y_index) references "Y" (index);
```

> > In the `insertUpdate` function (an actual real case I need to handle), the first commit adds two rows (2, 20) and (2, 21) to `XY` while the second simply removes the (2, 21) row (and does not add a second (2, 20) row). I would want to have two (2, 20) rows.
> 
> from a relational database perspective, there's no valid reason to have duplicate rows in a table (database relations are sets. dupes are meaningless). I'd ask what the actual problem you are trying to solve is, as proceeding using more traditional means will most definitely have a better end result.

Not exactly an unexpected answer. I appreciate you asking for more details to attempt to resolve the situation. I'll consult with colleagues first and may get back to you.

Thank you!

**Reply by zzzeek** [reply]
> > > ## Behavior and expectation
> > > `insert` adds two rows in `XY`, both (1, 10).
> > 
> > 
> > the ORM won't do this unless you had an additional primary key in the association table and you used the [association object pattern](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object)/
> 
> I'm not sure what you're saying. This is exactly what I'm observing: strictly using the code in the op and using the tables below (no primary key on `XY`), the `insert` function adds two rows to `XY`, both (1, 10). This is a postgres database, debian stable, sqlalchemy 1.4.39.

OK clarification, yes the ORM "does" it, but once it inserts those rows, it has no way to individually manipulate them.  that is, it's impossible in SQL to delete just one of the rows, and not the other.   Eager loaders, especially joinedload(), have no way to be aware of "dupes" like this either; joinedload() inherently requires deduplication of parent rows for it to work, so deduplication is already built into collection loading for joinedload.  for other loaders it's likely not defined.

So the pattern you have of adding the same object more than once is not supported even though it isn't actively checking for this mistake; this is why we recommend having a constraint on the table and/or using the set() collection type.

**Comment by yann-mabsilico** [comment, marked as answer]
Short update. I ended up adding an additional field (essentially a counter) to the association table and using an association object as suggested in the docs: https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object
